### DISCUSSION TITLE
Alpine Linux Container : ghcr.io/hellt/network-multitool : How to do Ethernet Bonding ?

### DISCUSSION THREAD

**Discussion by dharmbhai** [srl-labs/containerlab#366] [Q&A] [answered]
HI @hellt 
This is not directly related to containerlab, but to network-multitool container
I can console into it and do funky things.
However as a part of my MLAG testing, I wish to do ethernet bonding of 2 interfaces on network-multitool

Q1)Is it possible to do ethernet bonding on network-multitool in **docker** ? 

Q2) I am unable to find /etc/network/interfaces file in the network-multitool
So how do we do it in Alpine Linux/network-multitool
A google search reveals below
[](https://wiki.alpinelinux.org/wiki/Bonding)

Q3) Are there other containers that support ethernet bonding? I did a google search but nothing worthwhile came up

Please advice, else I shall be compelled to use cEOS-Lab for simulating End hosts (Servers) which is not ideal, but something I will have to live with.

```
bash-5.0# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
10: eth0@if11: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 02:42:ac:14:14:04 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 172.20.20.4/24 brd 172.20.20.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 2001:172:20:20::4/80 scope global nodad
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe14:1404/64 scope link
       valid_lft forever preferred_lft forever
19: eth2@if18: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 65000 qdisc noqueue state UP group default
    link/ether 26:d2:1c:3b:79:33 brd ff:ff:ff:ff:ff:ff link-netnsid 1
    inet6 fe80::24d2:1cff:fe3b:7933/64 scope link
       valid_lft forever preferred_lft forever
21: eth1@if20: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 65000 qdisc noqueue state UP group default
    link/ether 9a:50:9c:59:91:36 brd ff:ff:ff:ff:ff:ff link-netnsid 2
    inet6 fe80::9850:9cff:fe59:9136/64 scope link
       valid_lft forever preferred_lft forever
bash-5.0# uname -a
Linux client1 5.8.0-48-generic #54~20.04.1-Ubuntu SMP Sat Mar 20 13:40:25 UTC 2021 x86_64 Linux
bash-5.0# cat /etc/alpine-release
3.12.0
bash-5.0#
```

**Comment by hellt** [comment]
Hi @dharmbhai 
it should be possible to create bond/team interfaces. Since containers do not have an init system, you would not find a regular scripts in `/etc/interfaces`.
Instead you should use iproute2 utility that comes preinstalled with multitool container and can be installed to others containers with `apt update && apt install iproute2`.

Once installed, refer to the docs of ip utility on how to create bond/team interfaces - https://baturin.org/docs/iproute2/#ip-link-add-bond

some examples I [found](https://github.com/lxc/lxc/issues/2759) on internet:

```
root@bionic:~# ip link add name bond0 type bond
root@bionic:~# ip link add name tap0 type dummy
root@bionic:~# ip link add name tap1 type dummy
root@bionic:~# ip link set tap0 master bond0
root@bionic:~# ip link set tap1 master bond0
```

I haven't tried doing this myself, so if you get this working (I hope you will), this can be promoted to a doc enhancement in containerlab, so if you make this work, please update this thread with your findings =)

**Reply by dharmbhai** [reply]
Hi All
I have tried a lot of options, none worked, primarily due to my limited skills around docker networking.
I suppose, k8s could help achieve what we need, but that might be overkill for the containerlab

Overall, I understand we have to make the networking ready outside the container (using privilege mode and persistent mounts) for the container (Alpine Linux) to be able to bring up bond, but dont know how.

I request others to advice, in the interim, we can continue to use cEOS or for that matter any networking container to simulate the end host

**Reply by dharmbhai** [reply]
https://01.org/kubernetes/projects/bond-cni

Requires us to use K8s and implement multus

**Reply by hellt** [reply]
here is a working bond example I received (executed on an Ubuntu container)

```shell
# cat tools/ce-14/bond0.sh

 

# Create LACP based bond0 interface, LACP: fast, Load-Balancing: layer3+4

# Members: eth4, eth5

ip link add bond0 type bond mode 802.3ad

ip link set dev bond0 type bond xmit_hash_policy layer3+4

ip link set dev eth5 down

ip link set dev eth4 down

ip link set eth4 master bond0

ip link set eth5 master bond0

ip link set dev eth5 up

ip link set dev eth4 up

ip link set dev bond0 type bond lacp_rate fast

ip link set dev bond0 up

 

# Add VLAN based IP interface for IP-VRF/MAC-VRF network-instance

ip link add link bond0 name bond0.215 type vlan id 215

ip link set bond0.215 address 00:00:00:02:15:03

ip a add 192.168.215.2/24 dev bond0.215

ip link set dev bond0.215 up

 

# Optional static route, to reach remote subnet via IP-VRF

ip ro add 192.168.0.0/16 via 192.168.215.254

 

# Add VLAN based IP interface for MAC-VRF network-instance

ip link add link bond0 name bond0.120 type vlan id 120

ip link set bond0.120 address 00:00:00:01:20:14

ip a add 192.168.120.14/24 dev bond0.120

ip link set dev bond0.120 up

 

# bond0 status can be verified with

# cat /proc/net/bonding/bond0
```

**Reply by dharmbhai** [reply]
Hi @hellt 
These instructions work perfectly. Setting this discussion to answered
Thanks for all the help.

> here is a working bond example I received (executed on an Ubuntu container)
> 
> ```shell
> # cat tools/ce-14/bond0.sh
> 
>  
> 
> # Create LACP based bond0 interface, LACP: fast, Load-Balancing: layer3+4
> 
> # Members: eth4, eth5
> 
> ip link add bond0 type bond mode 802.3ad
> 
> ip link set dev bond0 type bond xmit_hash_policy layer3+4
> 
> ip link set dev eth5 down
> 
> ip link set dev eth4 down
> 
> ip link set eth4 master bond0
> 
> ip link set eth5 master bond0
> 
> ip link set dev eth5 up
> 
> ip link set dev eth4 up
> 
> ip link set dev bond0 type bond lacp_rate fast
> 
> ip link set dev bond0 up
> 
>  
> 
> # Add VLAN based IP interface for IP-VRF/MAC-VRF network-instance
> 
> ip link add link bond0 name bond0.215 type vlan id 215
> 
> ip link set bond0.215 address 00:00:00:02:15:03
> 
> ip a add 192.168.215.2/24 dev bond0.215
> 
> ip link set dev bond0.215 up
> 
>  
> 
> # Optional static route, to reach remote subnet via IP-VRF
> 
> ip ro add 192.168.0.0/16 via 192.168.215.254
> 
>  
> 
> # Add VLAN based IP interface for MAC-VRF network-instance
> 
> ip link add link bond0 name bond0.120 type vlan id 120
> 
> ip link set bond0.120 address 00:00:00:01:20:14
> 
> ip a add 192.168.120.14/24 dev bond0.120
> 
> ip link set dev bond0.120 up
> 
>  
> 
> # bond0 status can be verified with
> 
> # cat /proc/net/bonding/bond0
> ```

**Reply by dharmbhai** [reply]
Just a side note: Ubuntu container needs following to be done for ip link, ping, traceroute, etc to be available
docker container ubuntu image needs packages
apt update
apt upgrade
apt install apt-utils
echo 'debconf debconf/frontend select Noninteractive' | debconf-set-selections
apt install bridge-utils iproute2 inetutils-ping inetutils-traceroute

Once that is done, one can commit the container image and tag it the latest too, so that its available to others
docker commit c44db8091714 ubuntu:utils
sudo docker image tag ubuntu:utils  ubuntu:latest

**Comment by dharmbhai** [comment, marked as answer]
https://github.com/srl-labs/containerlab/discussions/366#discussioncomment-973920
